arkouda/apply.py

```diff
@@ -0,0 +1,99 @@
+from typeguard import typechecked
+from typing import Callable, Union, Optional
+
+from arkouda.client import generic_msg
+from arkouda.pdarrayclass import pdarray
+from arkouda.pdarraycreation import create_pdarray
+import cloudpickle
+import base64
+import numpy as np
+from arkouda.numpy.dtypes import dtype
+
+
+__all__ = [
+    "apply",
+]
+
+
+# TODO: it would be nice to typecheck that func takes and returns arr.dtype,
+# but that likely requires making pdarray generic over its dtype
+@typechecked
+def apply(
+    arr: pdarray,
+    func: Union[Callable, str],
+    result_dtype: Optional[Union[np.dtype, str]] = None,
+) -> pdarray:
+    """
+    Apply a python function to a pdarray. The function should take 1 argument
+    and return a new value. The function will then be called on each element in
+    the pdarray.
+
+    For example,
+    >>> apply(ak.array([1, 2, 3]), lambda x: x+1)
+
+    Or,
+    >>> import math
+    >>> arr = ak.randint(0, 10, 10_000)
+    >>> def times_pi(x):
+            return x*math.pi
+    >>> ak.apply(arr, times_pi, "float64")
+
+    Warning: This function is experimental and may not work as expected.
+    Known limitations:
+    - The use of some python modules inside of the function may crash the server.
+    Known modules know to crash the server are:
+        - numpy
```

Comment:
Based on your comment in the PR message, I was under the impression that this was more about using numpy from inside Jupyter. But I guess the issue is more general?

Reply:
Its not just numpy. I saw other modules crash it like matplotlib.

Reply:
The more general problem comes from cloudpickle. Its not clear if the problem is in how we use cloudpickle or in cloudpickle itself